fs/zonefs_zenfs.cc

```diff
@@ -140,16 +144,63 @@ IOStatus ZoneFsBackend::Finish(uint64_t start) {
   ret = truncate(filename.c_str(), zone_sz_);
   if (ret)
     return IOStatus::IOError("Zone finish failed: " + ErrorToString(errno));
+  PutZoneFile(start, O_WRONLY);
 
   return IOStatus::OK();
 }
 
-int ZoneFsBackend::OpenZone(uint64_t start, int flags) {
-  std::string filename = LBAToZoneFile(start);
-  return open(filename.c_str(), flags);
+std::shared_ptr<ZoneFsFile> ZoneFsBackend::GetZoneFile(uint64_t start,
+                                                       int flags) {
+  uint64_t zone_nr = start / zone_sz_;
+
+  if (flags & O_WRONLY) {
+    std::lock_guard<std::mutex> map_lock(zone_wr_fds_mtx_);
+    std::map<uint64_t, std::shared_ptr<ZoneFsFile>>::iterator open_fd =
+        wr_fds_.find(zone_nr);
+
+    if (open_fd == wr_fds_.end()) {
+      std::string filename = LBAToZoneFile(start);
+      int fd = open(filename.c_str(), flags);
+      if (fd == -1) return nullptr;
+
+      wr_fds_[zone_nr] = std::make_shared<ZoneFsFile>(fd);
+      return wr_fds_[zone_nr];
+    } else {
+      return open_fd->second;
+    }
+  } else {
+    std::lock_guard<std::mutex> map_lock(zone_rd_fds_mtx_);
+    std::pair<uint64_t, std::shared_ptr<ZoneFsFile>> *fd_cache =
```

Comment:
Maybe use a reference instead of a pointer here

Reply:
That would make sense. I have another change to switch to an LRU cache for all FDs, that will replace this, so I'll leave it as is for now.